mobilecoind-dev-faucet/src/worker.rs

```diff
@@ -406,20 +409,23 @@ struct WorkerTokenState {
     minimum_fee_value: u64,
     // The target value of UTXOS for this token id
     target_value: u64,
-    // The most recently known set of UTXOS for this token id
-    // When we get a new UTXO from mobilecoind, we track it using this cache.
-    // The tracker contains a one-shot channel that the other side can use to
-    // let us know what happens with this UTXO.
-    // UTXOs are added here at the same time they are queued. As long as a UTXO
-    // is in this cache, we won't requeue it, to avoid two threads spending
-    // the same UTXO concurrently.
-    known_utxos: HashMap<KeyImage, UtxoTracker>,
+    // Trackers for the UTXOs that we have queued.
+    // A tracker is created for a UTXO at the time that we get it from mobilecoind,
+    // and added here. The tracker contains a one-shot channel that the other side
+    // uses to let us know what happened with this UTXO.
+    // As long as a UTXO is in this list, we won't requeue it, until its tracker
+    // has resolved either in success or an error. We must avoid requeuing it
+    // to avoid handing it out to two different consumers who then conflict with
+    // each other.
+    queued_utxo_trackers: HashMap<KeyImage, UtxoTracker>,
     // The queue of UTXOs with the target value
     sender: async_channel::Sender<UtxoRecord>,
     // If we submit a rebalancing transaction, the response we can use to track it
     // Only one of these will be used at a time, and split txs cannot be submitted
     // while this is in-flight.
     in_flight_rebalancing_tx_state: Option<SubmitTxResponse>,
+    // Track the key images associated to utxos in the rebalancing tx
+    in_flight_rebalancing_key_images: HashSet<KeyImage>,
```

Comment:
I also added this as a minor improvement (I'm not sure it's really needed but it can't hurt)